### PR TITLE
feature: cli supports --memory-reservation

### LINKED ISSUE #2701: [feature request] pouch cli supports --memory-reservation
## Why you need it?

Cgroup memory soft limit is a general feature. Docker command supports it with "--memory-reservation". Pouch CRI supports it via "memory_reservation" in LinuxContainerResources.

## How it could be?

I'll be nice that pouch command supports --memory-reservation

### PR DESCRIPTION
**Pull request by KevinBetterQ**
Signed-off-by: KevinBetterQ <1093850932@qq.com>

### Ⅰ. Describe what this PR did
pouch command supports --memory-reservation

### Ⅱ. Does this pull request fix one issue?

fixes #2701 

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
added

### Ⅳ. Describe how to verify it
```
# pouch run -d --memory-reservation 4M busybox top
58369a469f0579234538717575d0338eb969b4009c6136ef32b678c432d8c072

# pouch inspect -f {{.HostConfig.MemoryReservation}} 5836
4194304

# cat /sys/fs/cgroup/memory/default/58369a469f0579234538717575d0338eb969b4009c6136ef32b678c432d8c072/memory.soft_limit_in_bytes
4194304
```

### CODE REVIEW

**Comment by CodeJuan** [comment]
Is it necessary to validate memory-reservation validateResource? I'm not sure there is a limitation of kernel that reservation should be less than 1/2 memory.
WDYT? @HusterWan @Ace-Tang

File: test/environment/env.go [hunk 1]
```diff
@@ -80,6 +80,11 @@ var (
 		return cgroupInfo.Memory.MemoryLimit
 	}
 
+	// IsMemoryReservationSupport checks if memory reservation cgroup is avaible
```
Comment (ZYecho) [inline test/environment/env.go]: avaible -> avaiable?

Comment (KevinBetterQ) [reply test/environment/env.go]: yes, it's a typo

**Comment by CodeJuan** [comment]
Hi @KevinBetterQ ,
 
I was wondering if you could help to develop a new feature `update memory-reservation by cli` at you convenience?

**Comment by KevinBetterQ** [comment]
> Hi @KevinBetterQ ,
> 
> I was wondering if you could help to develop a new feature `update memory-reservation by cli` at you convenience?

OK, I think it is a similar call in the code

File: apis/opts/memory_reservation.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package opts
+
+import units "github.com/docker/go-units"
+
+// ParseMemoryReservation parses the memory-reservation param of container.
+func ParseMemoryReservation(memoryReservation string) (int64, error) {
+	if memoryReservation == "" {
+		return 0, nil
+	}
+	result, err := units.RAMInBytes(memoryReservation)
```
Comment (HusterWan) [inline apis/opts/memory_reservation.go]: how about direct return `units.RAMInBytes(memoryReservation)`

Comment (KevinBetterQ) [reply apis/opts/memory_reservation.go]: modified

Comment (CodeJuan) [reply apis/opts/memory_reservation.go]: I'm afraid it is not a appropriate idea to return directly.
The upper function should handle the err of RAMInBytes and set memoryReservation to 0.

File: test/cli_run_memory_test.go [hunk 3]
```diff
@@ -164,6 +164,42 @@ func (suite *PouchRunMemorySuite) TestRunWithMemoryswappiness(c *check.C) {
 	c.Assert(out, check.Equals, memSwappiness)
 }
 
+// TestRunWithMemoryReservation is to verify the valid running container
+// with memory-reservation
+func (suite *PouchRunMemorySuite) TestRunWithMemoryReservation(c *check.C) {
```
[the remark this answers is not in the document]
Comment (KevinBetterQ) [reply test/cli_run_memory_test.go]: already in this function, just not adding a new function

File: apis/opts/memory_reservation.go (hunk 2 above)
Comment (KevinBetterQ) [reply apis/opts/memory_reservation.go]: if there is an error, the upper function will always return, will not care about the result.  and will return directly to the run execution failure, so there is no need to set this value.
Is that right?

File: test/cli_run_memory_test.go [hunk 4]
```diff
@@ -164,6 +164,59 @@ func (suite *PouchRunMemorySuite) TestRunWithMemoryswappiness(c *check.C) {
...
+	SkipIfFalse(c, environment.IsMemoryReservationSupport)
+
+	// test run with invalid memory reservation
+	cname := "TestRunWithMemoryReservationInvalid"
+	res := command.PouchRun("run", "-d",
+		"-m", "500m",
```
Comment (fuweid) [inline test/cli_run_memory_test.go]: could we use less memory here? According to the testing machine, I am not sure that it will impact the case.

**Pull request merged**